openedx/openedx-platform
# feat: emit COURSE_CATALOG_INFO_CHANGED signal on publish

This implements <https://github.com/openedx/edx-platform/issues/30682>.

## Depends on:
- [x]  https://github.com/openedx/openedx-events/issues/82 

## Description
Send new `COURSE_CATALOG_INFO_CHANGED` signal to the event bus whenever an appropriate course_published signal is sent.

## TODO

- [x] Figure out how to create CourseCatalogData from course_id
- [x] Abstract event bus library into setting (I couldn't figure out how to do this)
    - Deferred: openedx-events will handle configurable loading later in https://github.com/openedx/openedx-events/issues/87
- [x] Append environment type to topic name
    - I think we're deferring this.
- [x] Actually send to event bus
- [x] Copy all code over to draft_and_published (and hopefully figure out how to test it)
- [x] Undo accidental formatting changes
- [x] Make sure course keys are actually for courses and are not LibraryLocators

## Testing instructions

* In devstack, start Kafka and the control webapp: ``make dev.up.kafka-control-center`` and watch ``make dev.logs.kafka-control-center`` until server is up and happy (may take a few minutes; wait for http://localhost:9021/clusters to come up and show a healthy cluster)
* Add `SEND_CATALOG_INFO_SIGNAL = True` to `cms/envs/private.py` (enables feature, but also causes a server restart)
* Bring up studio with make `make dev.up.studio`
* In `make studio-shell`, run `make requirements` and then `pip install 'confluent_kafka[avro,schema-registry]'`
* Run `make dev.restart-devserver.studio` to make sure new requirements are picked up
* In studio, choose a course and go to Settings > Schedule & Details
* Change the pacing, start, or end date, and save
* In the Kafka control center (localhost:9021) navigate to Topics > course-catalog-info-changed > Messages
* Use Jump to Offset -> Offset 0 to see all messages
* You should see a message with the new course details
* If you keep this page open in the control center, you should be able to see all new messages as they come in. This should happen every time you change a field or create a new course run
